File: python/compress.py
import argparse
import os
import time
from concurrent.futures import ProcessPoolExecutor
from PIL import Image


def compress_image(input_path, output_path, quality):
    img = Image.open(input_path)
    output_file_path = os.path.join(output_path, os.path.basename(input_path))
    img.save(output_file_path, "WEBP", quality=quality)


def process_images(input_dir, output_dir, quality):
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    images = [
        os.path.join(input_dir, f)
        for f in os.listdir(input_dir)
        if f.endswith((".png", ".jpg", ".jpeg"))
    ]
    with ProcessPoolExecutor() as executor:
        for image_path in images:
            executor.submit(compress_image, image_path, output_dir, quality)


def main():
    parser = argparse.ArgumentParser(description="Image compression tool.")
    parser.add_argument(
        "--input-dir",
        required=True,
        help="Path to the input directory containing images.",
    )
    parser.add_argument(
        "--quality", type=int, required=True, help="WEBP compression quality."
    )
    parser.add_argument(
        "--output-dir",
        required=True,
        help="Path to the output directory for compressed images.",
    )

    args = parser.parse_args()

    start_time = time.time()
    process_images(args.input_dir, args.output_dir, args.quality)
    end_time = time.time()

    print(f"Total time taken: {end_time - start_time:.2f} seconds")


if __name__ == "__main__":
    main()

File: nodejs/compress.js
const fs = require('fs');
const path = require('path');
const { Worker } = require('worker_threads');
const { Command } = require('commander');
const os = require('os');

const program = new Command();

program
    .requiredOption('--input-dir <dir>', 'input directory containing images')
    .requiredOption('--quality <number>', 'WEBP compression quality', parseInt)
    .requiredOption('--output-dir <dir>', 'output directory for compressed images');

program.parse(process.argv);

const { inputDir, quality, outputDir } = program.opts();

if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
}

const files = fs.readdirSync(inputDir).filter(file => file.endsWith('.png') || file.endsWith('.jpg') || file.endsWith('.jpeg'));

// Limit the number of concurrent workers to the number of CPU cores
const maxWorkers = os.cpus().length;
console.log(`Workers: ${maxWorkers}`);
const promises = [];
let currentWorkers = 0;

const startTime = Date.now();

function processFile(file) {
    const inputPath = path.join(inputDir, file);
    const outputPath = path.join(outputDir, `${path.basename(file, path.extname(file))}.webp`);
    const worker = new Worker(path.join(__dirname, 'worker.js'), {
        workerData: { inputPath, outputPath, quality }
    });

    currentWorkers++;
    worker.on('message', (msg) => {
        // console.log(msg);
        currentWorkers--;
        if (files.length > 0) {
            processFile(files.shift());
        }
    });

    worker.on('exit', () => {
        if (currentWorkers === 0 && files.length === 0) {
            const endTime = Date.now();
            console.log(`Total time taken: ${(endTime - startTime) / 1000} seconds`);
        }
    });
}

// Start initial batch of workers
for (let i = 0; i < Math.min(maxWorkers, files.length); i++) {
    processFile(files.shift());
}
